### PR TITLE
Updates to vcfg rabbitmq shovel/federation jira_370

### PR DESCRIPTION
**Pull request by shwethanidd**
# Description

Fixes Jira tickets 370, 372 and 364. 
1. vcfg prompts for overwrite option if shovel/federation config already exists.
2. vcfg command now takes optional argument for max CSR retry attempts. Updated relevant documentation

### CODE REVIEW

**Comment by shwethanidd** [comment]
@sgilbride: I've made vcfg changes. Can you please review and test. Thanks.

File: volttron/utils/rmq_setup.py [hunk 1]
```diff
@@ -209,7 +205,7 @@ def _create_federation_setup(admin_user, admin_password, is_ssl, vhost, vhome):
                         rmq_user = upstream['federation-user']
                 else:
                     # certificates key not found in upstream config
-                    _log.debug("ERROR: certificates key not found in federation config. Cannot make connection to remote server without remote certificates")
+                    _log.info("ERROR: certificates key not found in federation config. Cannot make connection to remote server without remote certificates")
```
Comment (sgilbride) [inline volttron/utils/rmq_setup.py:208]: Could use _log.error(...)?

File: volttron/utils/rmq_setup.py [hunk 2]
```diff
@@ -276,7 +271,7 @@ def _create_shovel_setup(instance_name, local_host, port, vhost, vhome, is_ssl):
                         rmq_user = shovel['shovel-user']
                 else:
                     # destination key not found in shovel config
-                    _log.debug("ERROR: certificates key not found in shovel config. Cannot make connection to remote server without remote certificates")
+                    _log.info("ERROR: certificates key not found in shovel config. Cannot make connection to remote server without remote certificates")
```
Comment (sgilbride) [inline volttron/utils/rmq_setup.py:274]: Could use _log.error(...)?

File: volttron/utils/rmq_setup.py [hunk 3]
```diff
@@ -706,10 +701,10 @@ def setup_rabbitmq_volttron(setup_type, verbose=False, prompt=False, instance_na
                              rmq_config.volttron_home,
                              rmq_config.is_ssl)
     if invalid:
-        _log.error("Unknown option. Exiting....")
+        _log.error("Unknown option. Exiting XXX....")
```
Comment (sgilbride) [inline volttron/utils/rmq_setup.py:704]: What is the purpose of the XXX?

File: volttron/utils/rmq_setup.py [hunk 4]
```diff
@@ -573,7 +568,7 @@ def _verify_and_save_instance_ca(rmq_config, instance_ca_path, instance_ca_key):
 
 
 def setup_rabbitmq_volttron(setup_type, verbose=False, prompt=False, instance_name=None,
-                            rmq_conf_file=None, env=None):
+                            rmq_conf_file=None, max_retries=12, env=None):
```
Comment (sgilbride) [inline volttron/utils/rmq_setup.py:571]: Called in volttron/volttrontesting/fixtures/rmq_test_setup.py (line 189), but missing max_retries. Should it be added?

File: README.md [hunk 5]
```diff
@@ -150,7 +150,7 @@ You can deactivate the environment at any time by running `deactivate`.
 ##### 5. Create RabbitMQ setup for VOLTTRON:
 
 ```sh
-vcfg --rabbitmq single [optional path to rabbitmq_config.yml]
+vcfg rabbitmq single [--config optional path to rabbitmq_config.yml]
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply README.md:153]: rabbitmq single doesn't need max retry attempt. It is intended for federation/shovel. Max retry attempt for CSR operation with remote instance.

File: docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst [hunk 6]
```diff
@@ -625,13 +624,13 @@ agent. For such a request-response behavior, shovels need to be created on both
...
 
-   This says that Hplatform historian on v2 wants to make RPC call to DataMover agent on v1.
+   This says that platform historian on v2 wants to make RPC call to DataMover agent on v1.
 
    a. On v1, run below command to setup a shovel from v1 to v2.
 
   .. code-block:: bash
 
-     vcfg --rabbitmq shovel [optional path to rabbitmq_shovel_config.yml
+     vcfg rabbitmq shovel [optional path to rabbitmq_shovel_config.yml
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst:633]: fixed

File: docs/source/introduction/platform-install.rst [hunk 7]
```diff
@@ -293,7 +293,7 @@ Step 5 - Configure RabbitMQ setup for VOLTTRON
 
 .. code-block:: bash
 
-    vcfg --rabbitmq single [optional path to rabbitmq_config.yml]
+    vcfg rabbitmq single --config [optional path to rabbitmq_config.yml]
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/introduction/platform-install.rst:296]: rabbitmq single doesn't need max retry attempt. It is intended for federation/shovel. Max retry attempt for CSR operation with remote instance.

File: docs/source/platform-features/message-bus/rabbitmq/rabbitmq-ssl-auth.rst [hunk 8]
```diff
@@ -90,7 +90,7 @@ The parameters of interest for SSL based configuration are
 
 We can then configure the VOLTTRON instance to use SSL based authentication with the below command:
 
-    vcfg --rabbitmq single <optional path to rabbitmq_config.yml>
+    vcfg rabbitmq single --config [optional path to rabbitmq_config.yml]
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/platform-features/message-bus/rabbitmq/rabbitmq-ssl-auth.rst:93]: rabbitmq single doesn't need max retry attempt. It is intended for federation/shovel. Max retry attempt for CSR operation with remote instance.

File: volttron/utils/rmq_setup.py (hunk 1 above)
Comment (shwethanidd) [reply volttron/utils/rmq_setup.py:208]: done

File: volttron/utils/rmq_setup.py (hunk 2 above)
Comment (shwethanidd) [reply volttron/utils/rmq_setup.py:274]: done

File: volttron/utils/rmq_setup.py (hunk 4 above)
Comment (shwethanidd) [reply volttron/utils/rmq_setup.py:571]: Fixed in volttron/volttrontesting/fixtures/rmq_test_setup.py

File: volttron/utils/rmq_setup.py (hunk 3 above)
Comment (shwethanidd) [reply volttron/utils/rmq_setup.py:704]: fixed

File: docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst [hunk 9]
```diff
@@ -65,7 +65,7 @@ To setup federation on the VOLTTRON instance, run the following command on the d
 
 .. code-block:: bash
 
-    vcfg --rabbitmq federation [optional path to rabbitmq_federation_config.yml]
+    vcfg rabbitmq federation --config [optional path to rabbitmq_federation_config.yml] --max-tries [optional maximum CSR retry attempt]
```
[the remark this answers is not in the document]
Comment (shwethanidd) [reply docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst:68]: Typo. Fixed everywhere.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sgilbride) [reply README.md:153]: Sounds good. I would recommend moving config outside of the brackets for consistency: vcfg rabbitmq single --config [optional path to rabbitmq_config.yml]

[the remark this answers is not in the document]
Comment (craig8) [reply README.md]: I disagree with spencer...having it outside the bracket implies that you have to include --config whether you specify anything or not.

[the remark this answers is not in the document]
Comment (craig8) [reply README.md]: but it should be either one way or the other not both for sure... :)

**Comment by shwethanidd** [comment]
@schandrika Can you review this PR?

File: volttron/utils/rmq_mgmt.py [hunk 10]
```diff
@@ -766,7 +766,7 @@ def delete_multiplatform_parameter(self, component, parameter_name, vhost=None):
             config_file = os.path.join(vhome, 'rabbitmq_federation_config.yml')
             key = 'federation-upstream'
         config = read_config_file(config_file)
-        print("Removing certificate paths from the shovel config file. Please remove remote certificates manually "
+        print(f"Removing certificate paths from the {component} config file. Please remove remote certificates manually "
```
Comment (schandrika) [inline volttron/utils/rmq_mgmt.py:769]: Provide path to federation/shovel certs in vhome

File: volttron/utils/rmq_setup.py [hunk 11]
```diff
@@ -168,7 +168,48 @@ def write_env_file(rmq_config, conf_file, env=None):
...
+                    upstream['certificates']['remote_ca'] = ca_file
+                    # public cert
+                    upstream['certificates']['public_cert'] = cert_file
+                    # private_key
+                    upstream['certificates']['private_key'] = prvt_file
+                    update_needed = True
+                else:
+                    _log.error(
+                        f"ERROR: Couldn't get CSR certificates from remote server. Check with admin of remote instance."
+                        "\nContinuing with other configurations")
```
Comment (schandrika) [inline volttron/utils/rmq_setup.py:200]: When a csr is denied I get the message "continuing with other configurations" twice even though there is no more federation configuration in the file. File had only one entry. Looks like _get_federation_certs is returning True even when csr is denied and hence _create_federation_setup is called which throws "certificate key missing"

(volttron) volttron@volttron1:~/git/myvolttron$ vcfg rabbitmq federation

Your VOLTTRON_HOME currently set to: /home/volttron/fed_v1

Is this the volttron you are attempting to setup? [Y]:
Name of this volttron instance: [volttron1]:
rabbitmq_federation_config.yml exists in /home/volttron/fed_v1/rabbitmq_federation_config.yml Do you wish to use this file to configure federation [Y]:
Certificates not found. Requesting CSR from https://volttron2:8443
INFO: Request for signed certificate(CSR) has been denied by the remote instance administrator
ERROR: ERROR: Couldn't get CSR certificates from remote server. Check with admin of remote instance.
Continuing with other configurations
ERROR: ERROR: Certificates key missing in config.
Continuing with other configurations

File: volttron/utils/rmq_setup.py [hunk 12]
```diff
@@ -168,7 +168,48 @@ def write_env_file(rmq_config, conf_file, env=None):
...
+                    continue
```
Comment (schandrika) [inline volttron/utils/rmq_setup.py:201]: Also the error message here is different from the error message I get  when I do vcfg rabbitmq federation without a config file. Error message should be consistent. When no config file is used and csr is denied,  the error message clearly says that the csr was denied. I should get the same message when config file is used

When max retry is reached while using config file, I see a stack trace before the message "max retry attempt reached". When the same happens when running without config file there is no stack trace but just the message is displayed.  I like the latter 

**
Your VOLTTRON_HOME currently set to: /home/volttron/fed_v1

Is this the volttron you are attempting to setup? [Y]:
Name of this volttron instance: [volttron1]:
rabbitmq_federation_config.yml exists in /home/volttron/fed_v1/rabbitmq_federation_config.yml Do you wish to use this file to configure federation [Y]:
Certificates not found. Requesting CSR from https://volttron2:8443
INFO: Waiting for administrator to accept the CSR request.
Traceback (most recent call last):
  File "/home/volttron/git/myvolttron/env/bin/vcfg", line 11, in <module>
    load_entry_point('volttron', 'console_scripts', 'vcfg')()
  File "/home/volttron/git/myvolttron/volttron/platform/config.py", line 588, in _main
    main()
  File "/home/volttron/git/myvolttron/volttron/platform/instance_setup.py", line 1029, in main
    process_rmq_inputs(vars(args))
  File "/home/volttron/git/myvolttron/volttron/platform/instance_setup.py", line 967, in process_rmq_inputs
    setup_rabbitmq_volttron(args_dict['installation-type'], verbose, prompt=True, instance_name=instance_name, max_retries=args_dict['max_retries'])
  File "/home/volttron/git/myvolttron/volttron/utils/rmq_setup.py", line 748, in setup_rabbitmq_volttron
    success = _get_federation_certs(rmq_config.volttron_home)
  File "/home/volttron/git/myvolttron/volttron/utils/rmq_setup.py", line 187, in _get_federation_certs
    ca_file, cert_file, prvt_file = _request_csr(rmq_user, remote_addr, 'federation')
  File "/home/volttron/git/myvolttron/volttron/utils/rmq_setup.py", line 1222, in _request_csr
    raise TimeoutError("Maximum retry attempts for CSR reached. "
TimeoutError: Maximum retry attempts for CSR reached. Please retry command once administrator of remote VOLTTRON instance is ready to approve certificate signing request
(volttron) volttron@volttron1:~/git/myvolttron$ vcfg rabbitmq federation --max-retries 1

Your VOLTTRON_HOME currently set to: /home/volttron/fed_v1

Is this the volttron you are attempting to setup? [Y]:
Name of this volttron instance: [volttron1]:
rabbitmq_federation_config.yml exists in /home/volttron/fed_v1/rabbitmq_federation_config.yml Do you wish to use this file to configure federation [Y]: n
INFO: New input data will be used to overwrite existing /home/volttron/fed_v1/rabbitmq_federation_config.yml
Number of upstream servers to configure: [1]:
Hostname of the upstream server:  volttron2
Port of the upstream server:  [5671]:
Virtual host of the upstream server:  [volttron]:

Do you have certificates signed by remote CA?  [N]:
Path to remote web interface:  [https://volttron2:8443]:
INFO: Waiting for administrator to accept the CSR request.
ERROR: Maximum retry attempts for CSR reached. Please retry command once administrator of remote VOLTTRON instance is ready to approve certificate signing request
(volttron) volttron@volttron1:~/git/myvolttron$

File: volttron/utils/rmq_setup.py [hunk 13]
```diff
@@ -183,62 +224,94 @@ def _create_federation_setup(admin_user, admin_password, is_ssl, vhost, vhome):
...
+                                   vhost)
+
+            policy_name = 'volttron-federation'
+            policy_value = {"pattern": "^volttron",
+                            "definition": {"federation-upstream-set": "all"},
+                            "priority": 0,
+                            "apply-to": "exchanges"}
+            rmq_mgmt.set_policy(policy_name,
+                                policy_value,
+                                vhost)
```
Comment (schandrika) [inline volttron/utils/rmq_setup.py:262]: If setup was successful display message that it was successful. Else the last message before the command exits is "waiting for csr to be approved"

File: docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst [hunk 14]
```diff
@@ -165,8 +165,8 @@ upstream servers on the downstream server and make the VOLTTRON exchange
 
         .. code-block:: bash
 
-            vcfg --rabbitmq federation [optional path to rabbitmq_federation_config.yml
-            containing the details of the upstream hostname, port and vhost.]
+            vcfg rabbitmq federation [--config optional path to rabbitmq_federation_config.yml
+            containing the details of the upstream hostname, port and vhost.] [--max-retries optional maximum CSR retry attempt]
 
 
         Example configuration for federation is available
```
Comment (schandrika) [inline docs/source/deploying-volttron/multi-platform/multi-platform-rabbitmq-deployment.rst:172]: Remove steps to create user manually after running vcfg rabbitmq federation or shovel

**Review by schandrika — CHANGES_REQUESTED** [review]
1. When csr approval request times out display a timed out message instead of denied message.
2. If config file contains existing federation details display message that that particular federation entry is skipped and the code is proceeding to next federation entry
3. ERROR: Unable to create RabbitMQ user for the agent. Check if RabbitMQ broker is running - when I get this error the code should exit but it keeps proceeding to the next step. In documentation mention before running vcfg rabbitmq shovel/federation, user should shutdown volttron but make sure rabbitmq is running. They can use the start-rabbitmq script if the rabbit is in default path or can update the path in that script and run it

**Comment by schandrika** [comment]
vcfg is not prompting for overwrite when I have a config file in vhome. Please see commands below

**Attempt one - changed the user name in vhome config file:
------------------------------------------------------------------**
(volttron) volttron@volttron1:~/git/myvolttron$ pwd
/home/volttron/git/myvolttron
(volttron) volttron@volttron1:~/git/myvolttron$ diff rabbitmq_federation_config2.yml $VOLTTRON_HOME/rabbitmq_federation_config.yml
3c3
<     federation-user: volttron1.federation
---
>     federation-user: volttron1.federation.1
(volttron) volttron@volttron1:~/git/myvolttron$ vctl rabbitmq remove-federation-parameters upstream-volttron2-volttron
Removing certificate paths from the federation-upstream config file. Please remove remote certificates manually from the VOLTTRON_HOME folder if needed
names:'certificates'
No Federation Parameters Found ['upstream-volttron2-volttron']
(volttron) volttron@volttron1:~/git/myvolttron$ vcfg rabbitmq federation --config ./rabbitmq_federation_config2.yml --max-retries 1

Your VOLTTRON_HOME currently set to: /home/volttron/fed_v1

Is this the volttron you are attempting to setup? [Y]:
Name of this volttron instance: [volttron1]:
Certificates not found. Requesting CSR from https://volttron2:8443
(volttron) volttron@volttron1:~/git/myvolttron$ diff rabbitmq_federation_config2.yml $VOLTTRON_HOME/rabbitmq_federation_config.yml
2a3,6
>     certificates:
>       private_key: /home/volttron/fed_v1/certificates/private/volttron1.federation.pem
>       public_cert: /home/volttron/fed_v1/certificates/federation/volttron2.volttron1.federation.crt
>       remote_ca: /home/volttron/fed_v1/certificates/federation/volttron2_ca.crt
(volttron) volttron@volttron1:~/git/myvolttron$


**Attempt 2 : changed the upstream name in vhome config
----------------------------------------------------------------**

(volttron) volttron@volttron1:~/git/myvolttron$ diff rabbitmq_federation_config2.yml $VOLTTRON_HOME/rabbitmq_federation_config.yml
2c2,6
<   volttron2:
---
>   volttron3:
>     certificates:
>       private_key: /home/volttron/fed_v1/certificates/private/volttron1.federation.pem
>       public_cert: /home/volttron/fed_v1/certificates/federation/volttron2.volttron1.federation.crt
>       remote_ca: /home/volttron/fed_v1/certificates/federation/volttron2_ca.crt
(volttron) volttron@volttron1:~/git/myvolttron$ vcfg rabbitmq federation --config ./rabbitmq_federation_config2.yml --max-retries 1

Your VOLTTRON_HOME currently set to: /home/volttron/fed_v1

Is this the volttron you are attempting to setup? [Y]:
Name of this volttron instance: [volttron1]:
Certificates not found. Requesting CSR from https://volttron2:8443
(volttron) volttron@volttron1:~/git/myvolttron$ diff rabbitmq_federation_config2.yml $VOLTTRON_HOME/rabbitmq_federation_config.yml
2a3,6
>     certificates:
>       private_key: /home/volttron/fed_v1/certificates/private/volttron1.federation.pem
>       public_cert: /home/volttron/fed_v1/certificates/federation/volttron2.volttron1.federation.crt
>       remote_ca: /home/volttron/fed_v1/certificates/federation/volttron2_ca.crt

**Comment by schandrika** [comment]
I think we should delete certs when we delete federation link so that when you recreate a federation link you will always get it from upstream server - either the old one if it was not deleted or a new one through fresh csr approval. This way certs on downstream will match with what upstream will try to verify against

**Comment by shwethanidd** [comment]
Creating a new PR which has all the fixes

**Pull request closed without merging**